korangar/src/loaders/animation/mod.rs

```diff
@@ -0,0 +1,474 @@
+            // As the bottom part and left part will not be in the offset.
+            frames.iter_mut().for_each(|frame| {
+                let new_width = max_width;
+                let new_height = max_height;
+                for framepart in frame.frameparts.iter_mut() {
+                    framepart.offset.x += max_offset_x - min_offset_x + 2;
+                    framepart.offset.y += max_offset_y - min_offset_y + 2;
+                }
+                frame.offset.x = min_offset_x;
+                frame.offset.y = min_offset_y;
+                frame.offset.x += max_offset_x - min_offset_x + 2;
+                frame.offset.y += max_offset_y - min_offset_y + 2;
+                frame.size = Vector2::new(new_width, new_height);
+                frame.remove_offset.x = max_offset_x - min_offset_x + 2;
+                frame.remove_offset.y = max_offset_y - min_offset_y + 2;
+            });
+            animations.push(Animation { frames });
+        }
+        let animation_data = AnimationData {
+            delays: (&vec[0]).actions.delays.clone(),
+            animation_pair: vec,
+            animations,
+            entity_type,
+        };
+        let hash = match entity_type {
+            EntityType::Player => format!("0_{}_{}", entity_filename[0], entity_filename[1]),
+            EntityType::Monster => format!("1_{}", entity_filename[0]),
+            EntityType::Npc => format!("2_{}", entity_filename[0]),
+            _ => format!("3"),
+        };
+        self.cache.insert(hash.clone(), animation_data.clone());
+        Ok(animation_data)
+    }
+
+    pub fn get(
+        &mut self,
+        sprite_loader: &mut SpriteLoader,
+        action_loader: &mut ActionLoader,
+        entity_filename: Vec<String>,
+        entity_type: EntityType,
+    ) -> Result<AnimationData, LoadError> {
+        let hash = match entity_type {
+            EntityType::Player => format!("0_{}_{}", entity_filename[0], entity_filename[1]),
+            EntityType::Monster => format!("1_{}", entity_filename[0]),
+            EntityType::Npc => format!("2_{}", entity_filename[0]),
+            _ => format!("3"),
+        };
+
+        match self.cache.get(&hash) {
+            Some(animation_data) => Ok(animation_data.clone()),
+            None => self.load(sprite_loader, action_loader, entity_filename, entity_type),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct FramePart {
+    pub animation_index: i32,
+    pub sprite_number: i32,
+    pub offset: Vector2<i32>,
+    pub upleft: Vector2<i32>, // Is only used for internal calculation
+    pub size: Vector2<i32>,
+    pub mirror: bool,
+    pub angle: f32,
+    pub color: Color,
+}
+
+#[derive(Clone)]
+pub struct Frame {
+    pub offset: Vector2<i32>,
+    pub upleft: Vector2<i32>,
+    pub size: Vector2<i32>,
+    pub remove_offset: Vector2<i32>, // Is only used for final shift
+    pub frameparts: Vec<FramePart>,
+}
+
+impl Frame {
+    // This function generate new rame

```

Comment:
Typo. You can also use "///" to make a function documentation instead.

Reply:
Fixed.